File: mac/Sources/LlmIdeMac/Services/AutomationService.swift
// High-level automation service: orchestrates automatic memory capture,
// cleanup, contradiction detection, and graph regeneration by delegating to
// the MemoryService (Task 4) and GraphService (Task 5) service tiers.
//
// Mirrors the TS service layer (Task 6 of the storage/service split). The
// LLM-driven capture paths (`captureFromAgentTurn` / `captureFromUI`) are
// intentional no-ops today — extraction lands in a later phase — so this ships
// safe (does nothing) rather than guessing facts. Cleanup and contradiction
// detection, however, are fully functional and unit-tested.
//
// Deviations from the original task template (forced by the real Phase 1
// types / module contents so the code compiles with no breaking changes):
//   - The template's `AgentContext` is renamed `AgentTurnContext`. A major
//     domain type `AgentContext` already lives in
//     `Agent/Models/AgentTypes.swift` (used by CodeAssistantPanel and the
//     code-assist API) with a different shape (active project + indexed
//     repos); redeclaring it here would be an invalid redeclaration.
//     `AgentTurnContext` is descriptive and parallels `captureFromAgentTurn`.
//   - `ChatMemoryFact.timestamp` is `Int` milliseconds (Phase 1 contract), so
//     the cleanup age cutoff is computed in ms (Int), not Double seconds.
//   - `CleanupReport` / `ContradictionReport` cannot use tuples (tuples are
//     not `Codable`); their entries are small `Codable & Sendable` structs
//     (`RemovedFact`, `FactError`, `Contradiction`). Field access
//     (`.removed[0].reason`, `.kept`, `.contradictions`, counts) is unchanged.
//   - `detectContradictions` matches affirmative/negative phrases over the
//     FULL fact text with word boundaries (the template grouped by the first
//     whitespace token, which would never pair "This project uses …" with
//     "This project does not use …") and flags a pair only when the two facts
//     share a non-polarity subject word, generalising the template's
//     hard-coded `npm` check.

import Foundation

/// Automation service protocol for automatic memory operations.
///
/// Conforming types must be `Sendable` (the bundled impl is an actor) so a
/// single shared instance is safe to reuse across the app. The capture methods
/// are fire-and-forget safe — they never block the caller on LLM work today.
protocol AutomationService: Sendable {
    /// Capture durable facts from a completed agent turn. No-op today; LLM
    /// extraction lands in a later phase.
    func captureFromAgentTurn(context: AgentTurnContext) async throws

    /// Capture durable facts from a UI action. No-op today.
    func captureFromUI(action: UIAction) async throws

    /// Remove stale (too old) and invalid (failed validation) facts from the
    /// repo's chat memory, returning a report of what was removed, kept, and
    /// errored. Degrades gracefully — never crashes on read failure; returns
    /// an empty report instead.
    func cleanupStaleFacts(repoRoot: URL, olderThanDays: Int) async throws -> CleanupReport

    /// Detect facts that make contradictory claims about the same subject.
    func detectContradictions(repoRoot: URL) async throws -> ContradictionReport

    /// Mark the repo's graph for regeneration after documentation changes.
    func regenerateOnDocChange(repoRoot: URL) async throws

    /// Mark the repo's graph for regeneration after code changes.
    func regenerateOnCodeChange(repoRoot: URL) async throws
}

/// Context for a single agent turn to capture facts from. The `timestamp` is
/// `TimeInterval` (seconds since epoch) — the moment the turn completed.
struct AgentTurnContext: Codable, Sendable {
    let repoRoot: URL
    let userMessage: String
    let agentReply: String
    let timestamp: TimeInterval
}

/// UI action types that may yield durable facts.
enum UIAction: Codable, Sendable {
    case agentReply(reply: String)
    case fileViewed(file: URL)
    case commandExecuted(command: String)
}

/// A fact removed during cleanup, with the reason (e.g. `stale_age` or the
/// joined validation errors).
struct RemovedFact: Codable, Sendable {
    let fact: ChatMemoryFact
    let reason: String
}

/// A fact that errored during cleanup (validation threw), with the message.
/// The fact is kept in such cases — it could not be proven invalid.
struct FactError: Codable, Sendable {
    let fact: ChatMemoryFact
    let error: String
}

/// A pair of facts that contradict each other, with the reason.
struct Contradiction: Codable, Sendable {
    let fact1: ChatMemoryFact
    let fact2: ChatMemoryFact
    let reason: String
}

/// Report returned by `cleanupStaleFacts`.
struct CleanupReport: Codable, Sendable {
    let removed: [RemovedFact]
    let kept: [ChatMemoryFact]
    let errors: [FactError]
}

/// Report returned by `detectContradictions`.
struct ContradictionReport: Codable, Sendable {
    let contradictions: [Contradiction]
}

/// Automation service implementation. An actor for thread safety; delegates
/// memory I/O to `MemoryService` and graph regeneration to `GraphService`.
final actor AutomationServiceImpl: AutomationService {
    private let memoryService: MemoryService
    private let graphService: GraphService

    init(
        memoryService: MemoryService = MemoryServiceImpl(),
        graphService: GraphService = GraphServiceImpl()
    ) {
        self.memoryService = memoryService
        self.graphService = graphService
    }

    func captureFromAgentTurn(context: AgentTurnContext) async throws {
        // TODO: LLM extraction lands in a later phase. Ship as a no-op so
        // auto-capture is safe (never invents facts) until then.
    }

    func captureFromUI(action: UIAction) async throws {
        // TODO: UI-based capture lands in a later phase.
    }

    func cleanupStaleFacts(repoRoot: URL, olderThanDays: Int = 30) async throws -> CleanupReport {
        var removed: [RemovedFact] = []
        var kept: [ChatMemoryFact] = []
        var errors: [FactError] = []

        do {
            let facts = try await memoryService.readChatMemory(repoRoot: repoRoot)
            let nowMs = Int(Date().timeIntervalSince1970 * 1000)
            let cutoffMs = nowMs - olderThanDays * Self.msPerDay

            for fact in facts {
                // Age check (timestamps are ms since epoch).
                if fact.timestamp < cutoffMs {
                    removed.append(RemovedFact(fact: fact, reason: "stale_age"))
                    continue
                }

                // Validation check. If validation itself throws, record the
                // error and keep the fact (we couldn't prove it invalid).
                do {
                    let validation = try await memoryService.validateFact(repoRoot: repoRoot, fact: fact)
                    if !validation.valid {
                        // Prefer the per-check details (joined), then fall back
                        // to the summary reason, then a generic label — matching
                        // the TS-side cleanup reconciliation shape.
                        let reason = validation.details?.joined(separator: ", ")
                            ?? validation.reason
                            ?? "invalid"
                        removed.append(RemovedFact(fact: fact, reason: reason))
                        continue
                    }
                } catch {
                    errors.append(FactError(fact: fact, error: String(describing: error)))
                }

                kept.append(fact)
            }

            // Persist the survivors if anything was pruned.
            if !removed.isEmpty {
                try await memoryService.writeChatMemory(repoRoot: repoRoot, facts: kept)
            }
        } catch {
            // Graceful degradation — never crash; return the partial (likely
            // empty) report. MemoryService.readChatMemory already degrades to
            // [] so this catch is a last-resort guard.
            print("Cleanup failed: \(error)")
        }

        return CleanupReport(removed: removed, kept: kept, errors: errors)
    }

    func detectContradictions(repoRoot: URL) async throws -> ContradictionReport {
        let facts = try await memoryService.readChatMemory(repoRoot: repoRoot)
        var contradictions: [Contradiction] = []

        // Affirmative/negative phrase pairs that signal opposing claims.
        let opposites = [("uses", "does not use"),
                         ("requires", "does not require")]

        for (pos, neg) in opposites {
            // Tokens of the polarity phrases themselves, excluded from
            // subject matching so a pair isn't flagged only because both
            // facts contain "use" / "does".
            let excluded = Set((pos + " " + neg)
                .lowercased()
                .components(separatedBy: .whitespaces))
            let posFacts = facts.filter { Self.containsWord($0.text, pos) }
            let negFacts = facts.filter { Self.containsWord($0.text, neg) }

            for f1 in posFacts {
                let subjects1 = Self.subjectWords(f1.text, excluding: excluded)
                for f2 in negFacts {
                    let subjects2 = Self.subjectWords(f2.text, excluding: excluded)
                    // Flag only when both facts discuss the same subject, so
                    // unrelated positive/negative facts aren't paired.
                    guard !subjects1.isDisjoint(with: subjects2) else { continue }
                    contradictions.append(Contradiction(
                        fact1: f1,
                        fact2: f2,
                        reason: "Conflicting statements about \(neg)"))
                }
            }
        }

        return ContradictionReport(contradictions: contradictions)
    }

    func regenerateOnDocChange(repoRoot: URL) async throws {
        try await graphService.regenerateGraph(repoRoot: repoRoot)
    }

    func regenerateOnCodeChange(repoRoot: URL) async throws {
        try await graphService.regenerateGraph(repoRoot: repoRoot)
    }

    // MARK: - Helpers

    /// Milliseconds in one day.
    private static let msPerDay = 24 * 60 * 60 * 1000

    /// Common English stop words excluded from subject matching so two facts
    /// aren't paired merely because both say "this project …". `project` is
    /// included because it appears in nearly every template example sentence
    /// and carries no discriminative signal for contradiction detection.
    private static let stopWords: Set<String> = [
        "the", "this", "that", "these", "those", "a", "an",
        "is", "are", "was", "were", "be", "been", "being",
        "for", "to", "of", "in", "on", "with", "and", "or",
        "not", "does", "do", "did", "has", "have", "had",
        "it", "its", "we", "you", "they", "our", "their", "project"
    ]

    /// Whole-word, case-insensitive match so "uses" doesn't hit "houses".
    /// Handles multi-word phrases (e.g. "does not use") as well.
    private static func containsWord(_ text: String, _ word: String) -> Bool {
        guard let regex = try? NSRegularExpression(
            pattern: "\\b" + NSRegularExpression.escapedPattern(for: word) + "\\b",
            options: [.caseInsensitive]) else { return false }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return regex.firstMatch(in: text, options: [], range: range) != nil
    }

    /// Lowercased alphabetic subject tokens (length >= 3) of `text`, minus
    /// stop words and the polarity-phrase tokens.
    private static func subjectWords(_ text: String, excluding excluded: Set<String>) -> Set<String> {
        let tokens = text.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
        return Set(tokens.filter { token in
            token.count >= 3
                && token.allSatisfy { $0.isLetter }
                && !stopWords.contains(token)
                && !excluded.contains(token)
        })
    }
}

File: extension/graphkit/services/automation-service.ts
// AutomationService: automatic memory capture, cleanup, and graph regen hooks.
//
// Phase 2 service layer. Consumes MemoryService (Task 1) and GraphService
// (Task 2). Auto-capture methods (captureFromAgentTurn / captureFromUI) are
// intentional no-ops for now — full LLM extraction lands in Phase 4 — so they
// are safe to fire-and-forget from agent replies today. cleanupStaleFacts and
// detectContradictions carry real logic now.
//
// As with the sibling services, this file honors the SHIPPED Phase 1 surface
// and documents every place the original task brief drifted from it:
//
// 1. `validation.errors.join(', ')` (brief) -> `validation.details ?? ... .
//     The shipped `ValidationResult` type has `{ valid, reason?, details?,
//     contradicts? }` — there is NO `errors` array on a single-fact result
//     (that field lives on the all-facts `ValidationReport`). Calling
//     `.join` on `undefined` would throw, so the reason string is built from
//     `details` (falling back to `reason`, then 'invalid') exactly like
//     MemoryService.validateAllFacts does.
//
// 2. `detectContradictions` keyword lookup (brief) -> substring matching.
//     The brief indexed single words (`split(/\s+/)`) but queried multiword
//     phrases (`'does not use'`); a map keyed by single words can never
//     contain `'does not use'`, so `negFacts` was always empty and zero
//     contradictions were ever returned. Matching `pos`/`neg` as substrings
//     of each fact's full text makes the stated test pass and is the
//     intended behavior. Full LLM-based detection remains a Phase 4 task.
//
// 3. Timestamp round-trip + testability. The Phase 1 storage layer only
//     persists each fact's `text` (as `- text` lines); on read it rebuilds
//     every fact with `timestamp: Date.now()`, so a fact written "40 days
//     ago" comes back fresh and the age branch of cleanupStaleFacts could
//     never be exercised through the public API. To make that logic
//     unit-testable the dependencies are constructor-injectable (defaults
//     remain the real singletons, so production behavior is unchanged). This
//     is additive and non-breaking.
//
// Imports use `.ts` specifiers (not `.js`) so the module loads under
// `node --test --experimental-strip-types` (Node >= 20); the repo tsconfig
// sets `allowImportingTsExtensions: true`, so bundler/`tsc` accept these too.

import { memoryService as defaultMemoryService } from './memory-service.ts';
import { graphService as defaultGraphService } from './graph-service.ts';
import type { ChatMemoryFact, ValidationResult } from '../types/memory.ts';

/**
 * Narrow port of MemoryService used by this service. Defining a local
 * interface (rather than importing the whole class) keeps the dependency
 * surface explicit and makes the cleanup logic unit-testable with a stub.
 * The real `memoryService` singleton satisfies this structurally.
 */
interface MemoryPort {
  readChatMemory(repoRoot: URL): Promise<ChatMemoryFact[]>;
  writeChatMemory(repoRoot: URL, facts: ChatMemoryFact[]): Promise<void>;
  validateFact(repoRoot: URL, fact: ChatMemoryFact): Promise<ValidationResult>;
}

/**
 * Narrow port of GraphService used by this service.
 */
interface GraphPort {
  regenerateGraph(repoRoot: URL): Promise<void>;
}

/**
 * Context from an agent turn for memory capture.
 */
export interface AgentContext {
  repoRoot: URL;
  userMessage: string;
  agentReply: string;
  timestamp: number;
}

/**
 * UI action types for memory capture.
 */
export type UIAction =
  | { type: 'agentReply'; reply: string }
  | { type: 'fileViewed'; file: URL }
  | { type: 'commandExecuted'; command: string };

/**
 * Result of cleaning up stale facts.
 */
export interface CleanupReport {
  removed: Array<{ fact: ChatMemoryFact; reason: string }>;
  kept: ChatMemoryFact[];
  errors: Array<{ fact: ChatMemoryFact; error: string }>;
}

/**
 * Result of contradiction detection.
 */
export interface ContradictionReport {
  contradictions: Array<{ fact1: ChatMemoryFact; fact2: ChatMemoryFact; reason: string }>;
}

/**
 * AutomationService provides automatic memory capture and cleanup.
 *
 * Initially implements basic versions; full automation (LLM extraction,
 * richer contradiction detection) arrives in Phase 4.
 */
export class AutomationService {
  // Fields are declared explicitly (not as constructor parameter properties)
  // so the file runs under Node's default type-stripping (strip-only mode),
  // which does not support TS parameter properties. See storage/* for the
  // same convention.
  private memoryService: MemoryPort;
  private graphService: GraphPort;

  constructor(
    memoryService: MemoryPort = defaultMemoryService,
    graphService: GraphPort = defaultGraphService
  ) {
    this.memoryService = memoryService;
    this.graphService = graphService;
  }

  /**
   * Capture facts from an agent turn.
   *
   * Fire-and-forget safe: never throws (errors are logged), so it can be
   * awaited-or-not from the agent reply path without risking the reply.
   * TODO: Implement LLM extraction in Phase 4.
   */
  async captureFromAgentTurn(context: AgentContext): Promise<void> {
    try {
      // TODO: Use LLM to extract facts from conversation.
      // For now, this is a no-op to maintain safety.
      void context; // intentionally unused until Phase 4
    } catch (err) {
      console.error('Agent turn capture failed:', err);
      // Never fail the agent reply.
    }
  }

  /**
   * Capture facts from UI actions.
   *
   * Never throws. TODO: Implement UI hooks in Phase 4.
   */
  async captureFromUI(action: UIAction): Promise<void> {
    try {
      // TODO: Implement UI-based capture.
      // For now, this is a no-op.
      void action; // intentionally unused until Phase 4
    } catch (err) {
      console.error('UI action capture failed:', err);
    }
  }

  /**
   * Clean up stale facts (older than specified days) and facts that no longer
   * validate (e.g. referencing missing files). Returns a full report; never
   * throws (errors land in `report.errors` / are logged).
   */
  async cleanupStaleFacts(repoRoot: URL, olderThanDays = 30): Promise<CleanupReport> {
    const report: CleanupReport = {
      removed: [],
      kept: [],
      errors: []
    };

    try {
      const facts = await this.memoryService.readChatMemory(repoRoot);
      const cutoffTime = Date.now() - (olderThanDays * 24 * 60 * 60 * 1000);

      for (const fact of facts) {
        // Check age first.
        if (fact.timestamp < cutoffTime) {
          report.removed.push({ fact, reason: 'stale_age' });
          continue;
        }

        // Then validate file references / text length.
        try {
          const validation = await this.memoryService.validateFact(repoRoot, fact);
          if (!validation.valid) {
            // NOTE: ValidationResult carries detail in `details` (string), not
            // an `errors[]` array (see file header, drift note #1).
            const reason = validation.details ?? validation.reason ?? 'invalid';
            report.removed.push({ fact, reason });
            continue;
          }
        } catch (err) {
          report.errors.push({ fact, error: String(err) });
        }

        report.kept.push(fact);
      }

      // Write cleaned facts back if anything was removed.
      if (report.removed.length > 0) {
        await this.memoryService.writeChatMemory(repoRoot, report.kept);
      }
    } catch (err) {
      console.error('Cleanup failed:', err);
    }

    return report;
  }

  /**
   * Detect contradictory facts.
   *
   * Currently does simple substring-based detection for a small set of
   * positive/negative phrase pairs sharing a common subject token. Full
   * LLM-based contradiction detection lands in Phase 4. Never throws.
   */
  async detectContradictions(repoRoot: URL): Promise<ContradictionReport> {
    try {
      const facts = await this.memoryService.readChatMemory(repoRoot);
      const contradictions: Array<{ fact1: ChatMemoryFact; fact2: ChatMemoryFact; reason: string }> = [];

      // Match positive/negative phrase pairs as substrings of each fact's full
      // text (see file header, drift note #2 — single-word indexing can never
      // match a multiword phrase like 'does not use').
      const opposites: Array<[string, string]> = [
        ['uses', 'does not use'],
        ['requires', 'does not require']
      ];

      for (const [pos, neg] of opposites) {
        const posFacts = facts.filter((f) => f.text.toLowerCase().includes(pos));
        const negFacts = facts.filter((f) => f.text.toLowerCase().includes(neg));

        for (const f1 of posFacts) {
          for (const f2 of negFacts) {
            // Shared-subject guard retained from the brief's intent: only flag
            // a conflict when both statements are about the same subject token
            // (e.g. 'npm'). Keeps the stub from over-flagging.
            if (
              f1.text.toLowerCase().includes('npm') &&
              f2.text.toLowerCase().includes('npm')
            ) {
              contradictions.push({
                fact1: f1,
                fact2: f2,
                reason: `Conflicting statements about ${neg}`
              });
            }
          }
        }
      }

      return { contradictions };
    } catch (err) {
      // Graceful degradation — never crash the caller.
      console.error('Contradiction detection failed:', err);
      return { contradictions: [] };
    }
  }

  /**
   * Regenerate graph on doc change. Never throws (errors are logged).
   */
  async regenerateOnDocChange(repoRoot: URL): Promise<void> {
    try {
      await this.graphService.regenerateGraph(repoRoot);
    } catch (err) {
      console.error('Doc change regeneration failed:', err);
    }
  }

  /**
   * Regenerate graph on code change. Never throws (errors are logged).
   */
  async regenerateOnCodeChange(repoRoot: URL): Promise<void> {
    try {
      await this.graphService.regenerateGraph(repoRoot);
    } catch (err) {
      console.error('Code change regeneration failed:', err);
    }
  }
}

// Singleton instance (uses the real MemoryService / GraphService by default).
export const automationService = new AutomationService();
